src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Frameworks/SupportedTargetFrameworksProvider.cs

```diff
@@ -0,0 +1,93 @@
+    {
+        private readonly IProjectSubscriptionService _subscriptionService;
+
+        public SupportedTargetFrameworksProvider(
+            ConfiguredProject project,
+            IProjectSubscriptionService subscriptionService)
+            : base(project, synchronousDisposal: true, registerDataSource: false)
+        {
+            _subscriptionService = subscriptionService;
+        }
+
+        protected override IDisposable? LinkExternalInput(ITargetBlock<TOutput> targetBlock)
+        {
+            IProjectValueDataSource<IProjectSubscriptionUpdate> source = _subscriptionService.ProjectBuildRuleSource;
+
+            // Transform the changes from evaluation/design-time build -> restore data
+            DisposableValue<ISourceBlock<TOutput>> transformBlock = source.SourceBlock.TransformWithNoDelta(
+                update => update.Derive(Transform),
+                suppressVersionOnlyUpdates: false,
+                ruleNames: SupportedTargetFrameworkAlias.SchemaName);
+
+            // Set the link up so that we publish changes to target block
+            transformBlock.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
+
+            // Join the source blocks, so if they need to switch to UI thread to complete 
+            // and someone is blocked on us on the same thread, the call proceeds
+            JoinUpstreamDataSources(source);
+
+            return transformBlock;
+        }
+
+        private static ICollection<IEnumValue> Transform(IProjectSubscriptionUpdate input)
+        {
+                IProjectRuleSnapshot snapshot = input.CurrentState[SupportedTargetFrameworkAlias.SchemaName];
+
+                return snapshot.Items.Select(ToEnumValue)
+                                     .ToList();
+        }
+
+        private static IEnumValue ToEnumValue(KeyValuePair<string, IImmutableDictionary<string, string>> item)
+        {
+            return new PageEnumValue(new EnumValue()
+            {
+                // Example: <SupportedTargetFrameworkAlias  Include=".NETCoreApp, Version=v5.0"
+                //                                          TargetFramework="net5.0-windows"
+                //                                          DisplayName=".NET 5.0" />
+
+                Name = item.Value[SupportedTargetFrameworkAlias.TargetFrameworkProperty],
+                DisplayName = item.Value[SupportedTargetFrameworkAlias.DisplayNameProperty]
+            });
+        }
+
+        public Task<IDynamicEnumValuesGenerator> GetProviderAsync(IList<NameValuePair>? options)
+        {
+            return Task.FromResult<IDynamicEnumValuesGenerator>(this);
+        }
+
+        public async Task<ICollection<IEnumValue>> GetListedValuesAsync()
+        {
+            using (JoinableCollection.Join())
+            {
+                TOutput snapshot = await SourceBlock.ReceiveAsync();
```

Comment:
Does `SourceBlock` here refer to the same thing as the `source` local in `LinkExternalInput`?

Reply:
No, it's "target block" passed to LinkExternalInput. We're grabbing subscription data, transforming into a specific format, and pushing that onto the target block. That block is returned by SourceBlock.